### ISSUE TITLE
DNS Resolver rotate feature

### ISSUE THREAD

**Issue by TSDevHub** [hickory-dns/hickory-dns#2214]
**Is your feature request related to a problem? Please describe.**
Hello, I want to DNS cache a query which yields multiple A records and rotate through the A records.

I know you guys have https://docs.rs/hickory-resolver/0.24.1/hickory_resolver/config/struct.ResolverOpts.html#structfield.rotate feature, but upon inspecting your resolver source code, doesn't seem hickory resolver don't use it to rotate between cached A records for a given query, which would be useful for ip load balancing for services that use round-robin DNS. I also verified that with this rotate flag on, hickory resolver do not rotate between cached records, so if I am DNS caching, it just uses the same record even though in the IRU cache, hickory stores all the records for the query. This could be bad for overloading that particular ip.

Having the rotate feature work would make hickory resolver more adoptable for users trying to call service domain endpoint which use round-robin DNS for load balancing. Also might prevent users who enabled hickory resolver caching from unexpectedly bringing down their destination service hosts, due to surge in load in one ip due to caching.

**Describe the solution you'd like**
Allow hickory resolver to honor the rotate option, and round-robin rotate cached DNS query records. 

Suppose we made DNS query for domain www.myservice.com and it came up with these A records

```
1.2.3.4 IN A www.myservice.com
5.6.7.8 IN A www.myservice.com
9.10.11.12 IN A www.myservice.com
```

rotate would alternate fetching A records from DNS cache like so `1.2.3.4` -> `5.6.7.8` -> `9.10.11.12` -> back to -> `1.2.3.4` -> `5.6.7.8` -> so on, in round robin fashion, as long as cache is not expired

Also, would appreciate update in doc that this rotate flag currently is not implemented as it was misleading.

**Describe alternatives you've considered**
not use caching by fixing size = 0 for ip balancing, but.. not performant

**Additional context**
none

**Comment by bluejekyll** [comment]
yes, this would be a nice feature to add. I think it's also related to #1924, as this could be used in the forwarding hickory-dns server.

To make this accurate it would probably mean recording some stats on each RRSet and it's use in the cache, which I don't think we do today. That said, I think randomizing the order of the addresses is also something that the calling code can do. I've always been a little uncomfortable with changing RRSets from what was returned upstream, but I'm not a purest on this.

If you wanted to post something up, maybe even keeping it simple for now by just randomizing the order of the records in the result from the cache, that would be interesting to see.

**Comment by TSDevHub** [comment]
Hey Benjamin, thanks for your reply :)

Ah I didn't know hickory also provided DNS server capability, I was just using it as a resolver to call DNS server.

I think having a feature where you select a random record in the DNS cache if cache item has multiple records sounds like a great start. Perhaps hickory-resolver users would find that this is good enough for their use-case with short enough TTL. And later we can add round robin.

I am new to this repo and wonder how features are prioritized and released, what is next?

**Comment by bluejekyll** [comment]
we basically include things as they're ready and then publish in the next release. We're working on a new `0.25` release that will probably come once we have all of the related crates that use `ring` and `rustls` updated. So that will come as soon as some of our dependencies are updated.

**Comment by bluejekyll** [comment]
Something to consider when implementing this, perhaps it could be included only in the RRSet returned by the lookup function... i.e. that way it would make it something like `lookup_ip(...).await?.randomized().next()`

**Comment by TSDevHub** [comment]
I see, then I will keep an eye on this feature request. Thanks again!

**Comment by Xuanwo** [comment]
Hi, @TSDevHub, are you working on this now? I recently found that I need this feature and would be happy to collaborate on it.

**Comment by TSDevHub** [comment]
Hey @Xuanwo, I am currently not working on this and been waiting for the `0.25` release so it can picked up after. Good to see that more people want this feature :)

As for collaboration, I am uncertain if I would be able to collaborate with you at the moment so if you want to work on it now, please feel free to start

**Comment by aajtodd** [comment]
What is the intent of the [rotate](https://docs.rs/hickory-resolver/latest/hickory_resolver/config/struct.ResolverOpts.html#structfield.rotate) resolver option? Is it meant to work as described in this issue (i.e. if there are multiple cached entries, rotate through them on each invocation)? Looking at the code it doesn't seem to do much today, so I'm curious the original intent and what purpose it serves right now.

**Comment by djc** [comment]
I submitted #2841 to remove the `rotate` option for now, since, as you mentioned, it doesn't do anything today. Happy to review changes that bring back similar functionality.
